crates/client/settlement_client/src/starknet/mod.rs

```diff
@@ -0,0 +1,470 @@
+            l2_core_contract: config.l2_contract_address,
+            l1_block_metrics: config.l1_block_metrics,
+        })
+    }
+
+    async fn get_latest_block_number(&self) -> anyhow::Result<u64> {
+        let block_number = self.provider.block_number().await?;
+        Ok(block_number)
+    }
+
+    async fn get_last_event_block_number(&self) -> anyhow::Result<u64> {
+        let latest_block = self.get_latest_block_number().await?;
+        // If block on l2 is not greater than or equal to 6000 we will consider the last block to 0.
+        let last_block = if latest_block <= 6000 { 0 } else { 6000 };
+        let last_events = self
+            .get_events(
+                BlockId::Number(last_block),
+                BlockId::Number(latest_block),
+                self.l2_core_contract,
+                // taken from : https://github.com/keep-starknet-strange/piltover/blob/main/src/appchain.cairo#L102
+                vec![get_selector_from_name("LogStateUpdate")?],
+            )
+            .await?;
+
+        let last_update_state_event = last_events.last();
+        match last_update_state_event {
+            Some(event) => {
+                /*
+                GitHub Ref : https://github.com/keep-starknet-strange/piltover/blob/main/src/appchain.cairo#L101
+                Event description :
+                ------------------
+                #[derive(Drop, starknet::Event)]
+                struct LogStateUpdate {
+                    state_root: felt252,
+                    block_number: felt252,
+                    block_hash: felt252,
+                }
+                */
+                assert_eq!(event.data.len(), 3, "Event response invalid !!");
+                // Block number management in case of pending block number events.
+                match event.block_number {
+                    Some(block_number) => Ok(block_number),
+                    None => Ok(self.get_latest_block_number().await? + 1),
+                }
+            }
+            None => {
+                bail!("No event found")
+            }
+        }
+    }
+
+    async fn get_last_verified_block_number(&self) -> anyhow::Result<u64> {
+        let call_res = self
+            .provider
+            .call(
+                FunctionCall {
+                    contract_address: self.l2_core_contract,
+                    /*
+                    GitHub Ref : https://github.com/keep-starknet-strange/piltover/blob/main/src/state/component.cairo#L59
+                    Function Call response : (StateRoot, BlockNumber, BlockHash)
+                    */
+                    entry_point_selector: get_selector_from_name("get_state")?,
+                    calldata: vec![],
+                },
+                BlockId::Tag(BlockTag::Pending),
+            )
+            .await?;
+        assert_eq!(call_res.len(), 3, "Call response invalid !!");
+        // Block Number index in call response : 1
+        Ok(call_res[1].to_u64().unwrap())
```

Comment:
So here unwrap is justified Ig because we are already asserting the length and we know that the returned value can always be converted from Felt to u64. So this unwrap here is for Option<> type